store/model.go

```diff
@@ -0,0 +1,252 @@
+package store
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/alecthomas/jsonschema"
+	jsonpatch "github.com/evanphx/json-patch"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/textileio/go-eventstore"
+)
+
+var (
+	ErrNotFound   = errors.New("instance not found")
+	ErrReadonlyTx = errors.New("read only transaction")
+)
+
+type operationType string
+
+const (
+	upsert operationType = "upsert"
+	delete operationType = "delete"
+)
+
+type Model struct {
+	lock            sync.RWMutex
+	schema          *jsonschema.Schema
+	valueType       reflect.Type
+	datastore       ds.Datastore
+	dispatcher      *eventstore.Dispatcher
+	dispatcherToken eventstore.Token
+}
+
+func (m *Model) Read(f func(txn *Txn) error) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txn := &Txn{model: m, readonly: true, ops: make(map[ds.Key]operation)}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Model) Update(f func(txn *Txn) error) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txn := &Txn{model: m, ops: make(map[ds.Key]operation)}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return txn.Commit()
+}
+
+func (m *Model) FindByID(id string, v interface{}) error {
+	return m.Read(func(txn *Txn) error {
+		return txn.FindByID(id, v)
+	})
+}
+
+func (m *Model) Add(id string, v interface{}) error {
+	return m.Update(func(txn *Txn) error {
+		return txn.Add(id, v)
+	})
+}
+
+func (m *Model) Delete(id string) error {
+	return m.Update(func(txn *Txn) error {
+		return txn.Delete(id)
+	})
+}
+
+func (m *Model) Has(id string) (exists bool, err error) {
+	m.Read(func(txn *Txn) error {
+		exists, err = txn.Has(id)
+		return err
+	})
+	return
+}
+
+func (m *Model) Reduce(event eventstore.Event) error {
+	log.Debugf("reducer %s start", m.schema.Ref)
+	if event.Type() != m.schema.Ref {
+		log.Debugf("ignoring event from uninteresting type")
+		return nil
+	}
+	var op operation
+	if err := json.Unmarshal(event.Body(), &op); err != nil {
+		return err
+	}
+
+	key := ds.NewKey(event.EntityID())
+	switch op.Type {
+	case upsert:
+		value, err := m.datastore.Get(key)
+		if errors.Is(err, ds.ErrNotFound) {
+			if err = m.datastore.Put(key, op.JSONPatch); err != nil {
```

Comment:
This may seem odd. Considering is a `upsert` operation, if the value isn't found, is created.
May be better to separate `insert` from `update` operations, but since this json-patch operations is a temporal thing to see something working, made some kind of sense. If this operations would be separate, if we receive a `update` operation and the key isn't known, we should fail and kind of decide what to do with this unapplicable operation.